### ISSUE TITLE
Have jog mode behave more similarly to scratch mode

### ISSUE THREAD

**Issue by robbert-vdh** [mixxxdj/mixxx#11091]
### Feature Description

Having recently borrowed a DDJ-400 from a friend, I noticed that the jog wheel controls feel off compared to when I used it with Rekordbox at their place. Like most other controllers, there are three gestures involving the jog wheel:

1. Touching the top of the jog enables scratch mode, causing playback to stop and allowing you to directly move the playhead by rotating the jog wheel.
2. Rotating the wide of the jog wheel without touching the top is used for nudging and does the same thing, but at a slightly lower speed and without pausing the track (spinning it fast enough will cause the track to play backwards). 
3. Finally, pressing shift while rotating the jog wheel causes the playhead to jump through the track in 4 beat (or probably some time value if the track hasn't been analyzed yet) increments.

In Mixxx the scratching in vinyl mode from 1) feels great and responsive. But 2) and 3) feel very floaty. When you start rotating the jog wheel at a constant pace it takes almost two seconds of ramping up before the playhead also moves at the target pace, and when you stop rotating the jog again it also takes almost two seconds before it settles. As if there was a low-pass filter with a super low cutoff applied to the velocity of the jog wheel. This makes it very difficult to quickly move to the desired part of the track. Whereas I could jump to where I wanted to go within a second in Rekordbox thanks to its instant skipping in 4(?) beat increments, in Mixxx navigating around tracks this way takes me 10-20 seconds because it's so imprecise and because of a bug I'll get to next. And this smoothing makes nudging also very difficult, because if you don't anticipate for it you'll likely overshoot your alignment. That's probably also the reason why the nudging sensitivity is set super low on the DDJ-400 script. It takes me a bit over four full turns to shift the timing by one beat at 174 BPM.

I experimented a bit by changing the Shift+jog behavior in the [DDJ-400 script](https://github.com/mixxxdj/mixxx/blob/9965d55f6e00cd1982f78cf09ac6196887869487/res/controllers/Pioneer-DDJ-400-script.js#L475-L478) to use `engine.scratchTick()` with a much lower multiplier instead of the `engine.setValue(..., "jog", ...)` it's currently using. That, while not being the beat search behavior I'd prefer, is very responsive just like regular scratching and makes navigating through tracks much simpler. So having the generic jog mode behave more similar to the scratch mode (I haven't looked at any of the C++ source code yet) would greatly improve the responsiveness of using Mixxx with controllers.

And as for that bug I mentioned, the generic jog mode in the way it's currently used for skipping through tracks (with huge value modifiers) sometimes causes huge sudden drastic jumps. This usually happens when moving the jog wheel very slowly, or similarly at the end of a gesture when the jog wheel is slowing down while you're still pressing Shift. I already asked on Zulip, but do you want me to submit a separate bug report for that?

TL;DR: It feels as if there's a low-pass filter on the jog mode inputs causing everything to be floaty and imprecise compared to scratch mode, and there's a bug with the jog wheel where it causes huge sudden overshoots when rotating the jog at just the right pace.

**Comment by ronso0** [comment]
> When you start rotating the jog wheel at a constant pace it takes almost two seconds of ramping up before the playhead also moves at the target pace, and when you stop rotating the jog again it also takes almost two seconds before it settles.

Correct me if I'm wrong: isn't that configurable in  Preferences > Decks > Speed (tempo) ... options > Pitch bend behavior?

Btw TL;DR should be at the top :laughing:

**Comment by ronso0** [comment]
Re skipping beats, take a look at the beatjump controls, maybe that's what you're looking for
https://manual.mixxx.org/2.3/chapters/appendix/mixxx_controls.html#control-[ChannelN]-beatjump

**Comment by robbert-vdh** [comment]
> Correct me if I'm wrong: isn't that configurable in Preferences > Decks > Speed (tempo) ... options > Pitch bend behavior?

That doesn't affect this (that's presumably a low-pass filter for the pitch fader), and I also have that set to the default 'Abrupt jump' value.

But I just opened up the code that does something with the jog value and I immediately spotted this `Rotary` thing, which seems to implement a 25 element box filter: https://github.com/mixxxdj/mixxx/blob/9965d55f6e00cd1982f78cf09ac6196887869487/src/engine/controls/ratecontrol.cpp#L159-L162

That's definitely the issue. An FIR box filter sounds like a bad idea here, especially one that only receives a single value per buffer. I'm using a 512 sample period right now for debugging so at a sample rate of 48 kHz that means it's averaging values from the last quarter second. But I'm probably missing something because it feels more like two seconds. Does Mixxx use a different internal sample rate? That `Rotary` is also only used for the pitch fader and jog smoothing, and the vinyl scratch mode uses an entirely different filter. So presumably most of the problems with the jog controls can be solved by getting rid of the `Rotary` filter and replacing it by the same filter used for the `scratchTick` vinyl control function. Or can the two be merged There's probably a good reason why that hasn't been done yet.

**Comment by PatrickJosh** [comment]
I've experimented a bit and started by lowering the size of the buffer in the `Rotary` object from 25 to 2 in the following code:

https://github.com/mixxxdj/mixxx/blob/fef7062fd9775c907a9c10f9869b6ef1bc22e214/src/engine/controls/ratecontrol.cpp#L169

This made the jogwheel way snappier, but a bit to fast, so I lowered the jog sensitivity to 0.05 in the following code:

https://github.com/mixxxdj/mixxx/blob/fef7062fd9775c907a9c10f9869b6ef1bc22e214/src/engine/controls/ratecontrol.cpp#L363-L364

There's this suggestion in the comment of the second block that the value should be configurable. Maybe it'd be a good idea to make the first value configurable as well? This could be beneficial to all users who want a snappier response of their jogwheels or want to make them less sensitive. However, a global setting might cause problems for people who use multiple controllers. Is there a possibility to set such things per controller?

In case there are no problems with that or other ideas how to handle that, I would actually be willing to look into that and maybe propose a PR for this, either as a global option or per controller, if that is somehow feasible. But unfortunately, this still has to wait for some time because, currently, I have very limited time due to exams coming up.

**Comment by ronso0** [comment]
Thanks for looking into this issue!

> However, a global setting might cause problems for people who use multiple controllers. Is there a possibility to set such things per controller?

Not for the c++ code I'm afraid.
Though Mixxx now (2.5) has controller mapping settings so on the mapping level there could be some scaling applied.

**Comment by acolombier** [comment]
Note that this approach is already used for some settings in the [S4 Mk3](https://github.com/mixxxdj/mixxx/blob/main/res/controllers/Traktor-Kontrol-S4-MK3.js#L67)

**Comment by PatrickJosh** [comment]
> Though Mixxx now (2.5) has controller mapping settings so on the mapping level there could be some scaling applied.
> (edit in addition to some c++ refactoring?)

This would necessitate changes to each controller mapping though, wouldn't it? This seems rather impractical in my opinion. I think that this is something that people using other controllers might like to change too.

How about global settings that can be overwritten by settings on the mapping level? This way, people who are using controllers whose mappings don't support this feature yet can still use to global settings. In case this is sensible I'd first focus on implementing a global setting for this so that all people can benefit from these changes first.

It probably is rather out of scope to add some mechanism that enables settings per controller but without having to change each controller mapping, isn't it? 😅

**Comment by ghztomash** [comment]
I would like to report this issue as well, noticed it recently when playing with DDJ-FLX4 and DDJ-FLX2. 

The side jog wheel "nudge" just floats off and feels erratic, making beat matching very difficult.
On the other hand, the top platter scratching functionality on the jogwheel feels very responsive and direct.

I tried to change the [controller mapping alpha values](https://github.com/mixxxdj/mixxx/blob/26f9d320f0d10af45f9d9d9c1fff9830204fc16d/res/controllers/Pioneer-DDJ-FLX4-script.js#L180)
```js
PioneerDDJFLX4.alpha = 1.0/8;
```
But it didn't solve it.

The only thing that made it more usable was lowering the [m_pJogFilter](https://github.com/mixxxdj/mixxx/blob/26f9d320f0d10af45f9d9d9c1fff9830204fc16d/src/engine/controls/ratecontrol.cpp#L102) value to 6 or so.
```c++
m_pJogFilter(std::make_unique<Rotary>(25))
```

**Comment by ronso0** [comment]
Coming back to this this sounds much like the mod @Stefan-Franz wanted for his Hercules DJControl Instinct:
1. touching the top: always scratch
2. nudging the rim when playing: jog
3. nudging the rim when stopped: scratch 

We tried to solve 3. with `scratchEnable(..)` and `scratchTick()` it in the script and while it works, it's fragile IMO (or rather needs to be implemented very thoroughly) since the `play` control is 1 not just when playing (`play_latched`) but also when previewing hotcues.
And of course it's cumbersome/impossible to implement this for each and every mapping.

Otoh, we could simply ignore any `jog` input if `play` is 0 so users do scratching and positioning only with the top platter.
At least that feels like the least invasive fix for me 🤷‍♂️ 

Like, when play is set 0, clear the `jog` filter and/or return 0 in `RateControl::getJogFactor()`

What do yo think?

**Comment by PatrickJosh** [comment]
@ronso0 I don't quite understand how this is supposed to alleviate the problem while a track is playing, though. Of course, the problem is especially noticable when the track is stopped as you can experience this ‘floatiness’, but at least for me, this was an issue when a track was playing too, an from what it sounds like, for @ghztomash too, as they say that it makes beatmatching difficult. But maybe I just don't understand the proposed solution properly? 😅 

I am still committed to implement a global setting for the size of the rotary buffer as described [above](https://github.com/mixxxdj/mixxx/issues/11091#issuecomment-2657697329). Unfortunately, when I had a look at the code about a year ago, I did not find an easy solution how to implement it so that the value could be changed without having to restart Mixxx. And since then, uni and work have unfortunately been keeping me from investing more time in this.

**Comment by ronso0** [comment]
Yes, my proposal was just for the 'paused' situation.

Tbh I don't see the issue with `jog` when the track plays. I can tell for sure because I often do nudge with key unlocked, meaning I _hear_ when the track settles at normal speed which takes around .5 - .8 sec.
Fwiw I use a Traktor S4 Mk3 with a jog factor of 0.4, samplerate is 44. kHz.

Which samplerate do you use?
There's a [comment in RateControl suggesting the filter behavior depends on the samplerate](https://github.com/mixxxdj/mixxx/blob/05ee404a9a09b49b3c90a7a12d86806f83e403ed/src/engine/controls/ratecontrol.cpp#L101)

Have you tried adjusting the [`bendScale` const](https://github.com/mixxxdj/mixxx/blob/26f9d320f0d10af45f9d9d9c1fff9830204fc16d/res/controllers/Pioneer-DDJ-FLX4-script.js#L185) in the FLX4 mapping?


_____
> I tried to change the [controller mapping alpha values](https://github.com/mixxxdj/mixxx/blob/26f9d320f0d10af45f9d9d9c1fff9830204fc16d/res/controllers/Pioneer-DDJ-FLX4-script.js#L180)

The alpha / beta values are used to setup the scratch engine, unrelated to jog.

**Comment by ghztomash** [comment]
> I am still committed to implement a global setting for the size of the rotary buffer as described above. 

I do have this implemented on a branch if you would like to test it.

> I hear when the track settles at normal speed which takes around .5 - .8 sec.

I made some experiments with a 1000hz sine wave, and comparisons with XDJ-700 and FLX2, I can share the audio recordings in the following days, (the hardest part is is consistently nudjing the jog-wheel with the same force 😅)

For precise adjustments I would personally expect it to settle immediately when the jog-wheel stops physically spinning. And would expect a more dramatic spin down only after nudging the jog wheel with a lot of force, so that it keeps spinning and eventually slows down under it's own momentum. At the moment it feels a little bit like a turntable with a broken motor (very low torque).

I wonder how controller specific this is though.
